tests/EdgeToEdge/PayPalRequestLoggerTest.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+namespace WMDE\Fundraising\Frontend\Tests\EdgeToEdge;
+
+use org\bovigo\vfs\vfsStream;
+use org\bovigo\vfs\vfsStreamDirectory;
+use Symfony\Bundle\FrameworkBundle\KernelBrowser;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use WMDE\Fundraising\Frontend\App\EventHandlers\PayPalRequestLogger;
+
+/**
+ * @covers \WMDE\Fundraising\Frontend\App\EventHandlers\PayPalRequestLogger
+ */
+class PayPalRequestLoggerTest extends WebRouteTestCase {
+
+	private array $paypalRoutes = [
+		'handle_paypal_payment_notification',
+		'handle_paypal_membership_fee_payments'
+	];
+
+	private const LOG_DIR = 'log';
+	private const LOG_FILENAME = 'test.csv';
+
+	private vfsStreamDirectory $filesystem;
+	private KernelBrowser $client;
+	private string $filePath;
+
+	public function setUp(): void {
+		$this->filesystem = vfsStream::setup( self::LOG_DIR );
+		$this->filePath = vfsStream::url( self::LOG_DIR . '/' . self::LOG_FILENAME );
+
+		/** @var KernelBrowser $client */
+		$client = $this->createClient();
+		$this->client = $client;
+		self::$container->set(
+			PayPalRequestLogger::class,
+			new PayPalRequestLogger( $this->filePath, $this->paypalRoutes )
+		);
+	}
+
+	public function testWhenPostedPayPalData_createsLogFile() {
+		$route = self::$container->get( 'router' )->generate(
+			$this->paypalRoutes[0],
+			[],
+			UrlGeneratorInterface::RELATIVE_PATH
+		);
+		$payPalData = $this->validPayPalData();
+		$expectedContents = implode( ', ', $payPalData ) . PHP_EOL;
```

Comment:
here, and in `testWhenPostedPayPalDataAndFileExists_addsDataToLogFile` the test is repeating code from the implementation. This is a code smell IMHO. Tests should - even at the risk of being a bit WET and brittle -  explicitly spell out the expected data as strings instead of constructing these strings.
Appending `PHP_EOL` is fine, but I'd rather not use `implode`.